pyravendb/store/document_session.py

```diff
@@ -302,7 +302,12 @@ def store(self, entity, key="", change_vector=""):
             raise ValueError("None entity value is invalid")
 
         force_concurrency_check = self._get_concurrency_check_mode(entity, key, change_vector)
-        if entity in self._documents_by_entity:
+        try:
+            is_stored = entity in self._documents_by_entity
```

Comment:
Why do we need to support it like that? Why not use the object pointer instead of `__hash__` ?

Reply:
Python `dict()` uses `__hash__()` and `@dataclass` classes do not have it by default. There are [means](https://stackoverflow.com/questions/52390576/how-can-i-make-a-python-dataclass-hashable-without-making-them-immutable) to change that, one of them being implementing the `__hash__()` method. The purpose of this change is related to this [issue](https://groups.google.com/g/ravendb/c/RsJ5TEpCV78/m/lA_EAtAvAgAJ), pointing out an unclear error message.

Reply:
Easier to tell them about `unsafe_hash=True`, no?

Reply:
I believe it's unsafe. Hash changes if the content of the entity changes it seems, so the document won't be found again, if I'm not mistaken?

```
>>> @dataclass(unsafe_hash=True)
... class C:
...     a: int
...
>>> obj1=C(a=1)
>>> hash(obj1)
-6644214454873602895
>>> obj1.a=2
>>> hash(obj1)
6909455589863252355
>>>
...
>>> d={}
>>> d[obj1]="test"
>>> obj1
C(a=2)
>>> hash(obj1)
6909455589863252355
>>> obj1.a=3
>>> hash(obj1)
-5029647727744300836
>>> d[obj1]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: C(a=3)
>>>
```

Not sure what should be best here https://docs.python.org/3/library/dataclasses.html#dataclasses.dataclass
I'd say: avoid using unhashable types as entities.

How does it work for `record` types in C#? Haven't tried them out yet.

Reply:
Implementing `__hash__()` yourself won't help here neither.

Reply:
what we want to do is to have that done based on the `id()` of the object.
We want reference equality here. Instead of using the object directly, we can have a wrapper that would use the internal `id()` for the hash, no?

Reply:
```
from dataclasses import dataclass


class RefEq(object):
    def __init__(self, ref):
        self.ref = ref

    def __eq__(self, other):
        if id(self.ref) == id(other):
            return true
        if not isinstance(other, RefEq):
            return false
        return id(self.ref) == id(other.ref)
		

    def __hash__(self):
        return id(self.ref)


@dataclass()
class C:
    a: int


obj1 = C(a=1)
re = RefEq(obj1)

print(re.__hash__())
obj1.a = 2
print(re.__hash__())

print(re == RefEq(obj1))
```

Reply:
Maybe we could change `_documents_by_entity` to `_documents_by_entity_object_id` ?

Reply:
ah you want to wrap the entities objects instead. Sure works too.

Reply:
Using just the `id()` isn't a good idea, we may end up reusing the same pointer, which will be consuing.
With the wrapper, we ensure that we keep the id() busy

Reply:
Prepared PR to be merged, created new ticket for this issue - https://issues.hibernatingrhinos.com/issue/RDBC-471

Reply:
https://ayende.com/blog/194529-A/reference-equality-for-dictionaries-in-python?key=5b04d4f028f342cf9eb39a8c96e1523f